feat: Implement AI-powered Common Issues

chuporustum opened the pull request:
## 🎯 **Overview**

This PR introduces an **AI-powered Common Issues** system that intelligently categorizes and groups support conversations. The system combines automatic AI categorization with manual management capabilities, enabling support teams to efficiently handle similar issues at scale. This addresses the need for intelligent conversation organization as outlined in issue #483.

## 🏗 **Architecture Overview**

```mermaid
graph TB
    A[New Conversation] --> B[AI Categorization Job]
    B --> C{Match Found?}
    C -->|Yes| D[Auto-assign to Issue Group]
    C -->|No| E[Remain Unassigned]
    
    F[Support Agent] --> G[Manual Issue Creation]
    G --> H[Issue Group Database]
    
    I[Common Issues Page] --> J[View Grouped Issues]
    J --> K[Search & Sort]
    J --> L[Pin/Bookmark Groups]
    J --> M[Volume-based Badges]
    
    N[Conversation View] --> O[Issue Assignment Dropdown]
    O --> P{Select Action}
    P -->|Assign| Q[Update Group]
    P -->|Clear| R[Remove Assignment]
    
    S[Real-time Updates] --> T[WebSocket Events]
    T --> U[UI Refresh]
```

## 🚀 **Key Features**

### **AI-Powered Categorization**
- ✅ **Automatic Assignment**: New conversations are automatically categorized using GPT-4o-mini
- ✅ **Vector Similarity Search**: Uses embeddings to find the best matching issue group
- ✅ **Confidence Scoring**: Only assigns conversations when confidence is high
- ✅ **Real-time Processing**: Triggered immediately when new messages arrive

### **Common Issues Dashboard**
- ✅ **Elegant Card UI**: Modern stacked card effect with hover animations
- ✅ **Volume Indicators**: Color-coded badges showing ticket volume (red for high, gray for normal)
- ✅ **VIP Tracking**: Special indicators for issues affecting VIP users
- ✅ **Search & Sort**: Full-text search and sorting by frequency or recency
- ✅ **Bookmarking**: Pin frequently accessed issue groups for quick access

### **Manual Management**
- ✅ **Settings Page**: Create, edit, and delete issue groups manually
- ✅ **Bulk Actions**: Close all conversations in an issue group at once
- ✅ **Real-time Counts**: Live conversation counts per issue group
- ✅ **Conversation Assignment**: Manually assign/unassign conversations via dropdown

### **Navigation & Integration**
- ✅ **Sidebar Integration**: Pinned issues appear in the main sidebar
- ✅ **Deep Linking**: Click any issue group to see filtered conversations
- ✅ **Zero State**: Clean empty state with helpful guidance
- ✅ **Responsive Design**: Works seamlessly on all screen sizes

## 📊 **User Workflows**

### **Automatic AI Categorization Flow**
1. Customer sends a new support message
2. System generates embeddings for the conversation
3. AI analyzes content and searches for matching issue groups
4. If a high-confidence match is found, conversation is auto-assigned
5. Issue group badges update in real-time

### **Manual Issue Management**
1. Navigate to **Settings → Common Issues**
2. Create new issue groups with title and description
3. AI generates embeddings for better matching
4. Monitor conversation counts and volume
5. Delete unused groups with confirmation

### **Common Issues Dashboard**
1. Access via **Common Issues** in the sidebar
2. View all issue groups with volume indicators
3. Search for specific issues or sort by frequency/recency
4. Bookmark frequently used groups
5. Click any group to see all related conversations

## 🎨 **UI/UX Improvements**

### **Visual Design**
- Subtle stacked card effect for visual depth
- Theme-aware color system (no hard-coded colors)
- Consistent typography and spacing
- Smooth hover animations and transitions
- Responsive grid layout

### **Badge System**
- **Red badge**: High volume (10+ tickets)
- **Gray badge**: Normal volume (1-9 tickets)
- **Time-based display**: Shows "today", "this week", or "this month"
- **VIP indicator**: Yellow-bordered badge for VIP users

## 📱 **Screenshots & Demos**

### **Common Issues Dashboard**

https://github.com/user-attachments/assets/76c27cf6-5d79-4999-8af7-80000d7767e6

### **Manual Assignment Flow**

https://github.com/user-attachments/assets/1d235bdd-bed1-43f1-b4db-57a2bcba8117

### **Bookmarked Issues in Sidebar**

<img width="261" height="437" alt="Screenshot 2025-07-16 at 9 55 16 PM" src="https://github.com/user-attachments/assets/1539f697-0042-4532-8b6c-e7571f480ca9" />

### **Zero State**

<img width="1323" height="666" alt="common-issues-0-state" src="https://github.com/user-attachments/assets/23d695ae-7d16-4461-b813-c8fe2cf8ce2f" />

## 🔧 **Technical Details**

### **Database Changes**
- Added `issue_groups` table with embedding support
- Added `issueGroupId` to conversations table
- Added `user_pinned_issue_groups` for bookmarking
- Includes proper indexes for performance

## 🧪 **Testing**
- Comprehensive unit tests for AI categorization logic
- Integration tests for TRPC endpoints
- E2E tests for user workflows
- Manual testing across different screen sizes

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced "Common Issues" (issue groups) for grouping, assigning, filtering, and managing conversations by common problems.
  * Added dashboard page for browsing, searching, sorting, pinning, and unpinning issue groups with real-time updates and detailed metrics.
  * Enabled assignment and removal of conversations to/from issue groups via a new sidebar section and assign button component.
  * Added settings page to create, edit, and delete common issue groups.
  * Provided filtering of conversations by issue group in conversation lists and filters.
  * Added "Pinned Issues" section in the sidebar with quick access to frequently used issue groups.

* **Bug Fixes**
  * Improved localStorage handling to prevent errors during server-side rendering.

* **Chores**
  * Updated database schema with new tables, columns, and indexes to support issue groups and pinning.
  * Added background jobs for AI-powered automatic categorization of conversations into issue groups.
  * Integrated issue groups management into API routes and real-time event channels.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

chuporustum commented:
Marked **ready for review** so that AI reviews start.

Adarsh9977 commented:
@chuporustum Great architecture, and very nice explanation of your changes. If possible can you please add any video of your applied changes.

binary-koan commented:
Thanks for building this out! My first impression is that this has the same problems as the last time we tried clustering - "Billing and account access issues" is not a useful category for support staff to go through since it covers a huge range of issues (the level of specificity we want is things like "Can't get 2FA code to login" for example, which makes sense for an agent to tackle all at once and could even be dealt with in bulk)

The UI looks good though, will take a proper look through this when I have time and think about what approach we should take

chuporustum commented:
I modified the prompts a bit and increased the cosine similarity threshold to try out something. Was able to see more specific clusters (from 9 to 7 previously) with more specific naming

[Gist with results](https://gist.github.com/chuporustum/b8fa840616c00f8a27a293e824962208)

I think we can play around a bit with the same although not going to spend more time on this until your review as there can be other approaches also we could consider.

slavingia commented:
We shouldn't be using vector embedding but having an LLM categorize each new ticket that's coming in (probably with some delay as it'll take time for the conversation progress to have enough detail to tag).

binary-koan commented:
> For the reasons above the issue groups will be provided by humans, and be a single long list of 100+ issues. They should be very specific.

To expand on this, we should have

* A way to create these manually with a description string
* An action on the conversation page to "set common issue" which lets you pick or create an issue
* A job to categorize new conversations which prompts AI with something like "These are our issue groups. Pick the one that best suits this message content, or nothing if there is no good match:"

> Screenshots of the settings page where common issues can be CRUD would be great.

IMO we can just have common issues CRUD on the same page rather than separate settings

---

One other request from a brief look at the code: we don't need join tables for `conversationIssueGroups` and `userIssueGroups` IMO, for conversations we can link to a single issue and for users we can use a jsonb column (we deliberately don't use foreign keys anyway)

chuporustum commented:
> > > For the reasons above the issue groups will be provided by humans, and be a single long list of 100+ issues. They should be very specific.
> > 
> > 
> > To expand on this, we should have
> > 
> > * A way to create these manually with a description string
> > * An action on the conversation page to "set common issue" which lets you pick or create an issue
> > * A job to categorize new conversations which prompts AI with something like "These are our issue groups. Pick the one that best suits this message content, or nothing if there is no good match:"
> > 
> > > Screenshots of the settings page where common issues can be CRUD would be great.
> > 
> > 
> > IMO we can just have common issues CRUD on the same page rather than separate settings
> > One other request from a brief look at the code: we don't need join tables for `conversationIssueGroups` and `userIssueGroups` IMO, for conversations we can link to a single issue and for users we can use a jsonb column (we deliberately don't use foreign keys anyway)
> 
> @slavingia @binary-koan The CRUD functionality of common Issues is already created with proper schema in #698. If you allow I will modify that keywords with description and the AI job will be done in this one. Because CRUD is already tested and working on that. wdyt?

Hey, thanks! It's fine, I have already made the changes and about to push in this PR.

chuporustum commented:
Fixing few other errors (+ the comments from coderabbit) but this is how it looks like right now

<img width="1387" alt="Screenshot 2025-07-09 at 12 55 53 PM" src="https://github.com/user-attachments/assets/b00822d6-f79b-48cb-8f7f-f777bd6f979a" />

slavingia commented:
See our existing knowledge base UX for what I'm thinking.

slavingia commented:
Sweet, matching that UI and not using a modal, but just having it inline with autosave would be great, something like this:

<img width="367" alt="image" src="https://github.com/user-attachments/assets/6d878b88-1906-4a8d-a78a-5909aad24322" />

(UX wise, less UI wise)

slavingia commented:
This is simple enough that there should be a button from the common issues page to "Edit common issues" as well, and maybe that can live within a modal for quick updates?

chuporustum commented:
@slavingia @binary-koan possible to get some initial feedback on the UI now? (recordings in PR desc)

If this looks okay, I can wrap up the AI generation part today and we can move towards closure

slavingia commented:
I don't see any of the common issues page itself, settings etc look good

slavingia commented:
Could you try to get it pixel perfect to the design mock?

Font sizes, weight, icons, hover, etc

binary-koan commented:
> Please check the search bar placement in the demo video below

I'm fine with that placement, but we should use `rounded-full` for consistency

> Not sure what these 2 icons are doing so requesting some explaination

I would just remove all three tbh, I don't see a use case for share

> Not sure of the grey color being used in New tickets tag

Just `muted` I think - we should stick to the theme palette to keep Helper flexible

---

Quick things looking at the screencast

- The stack should only show on hover, and shouldn't have the blue border
- Is the VIP indicator implemented?

chuporustum commented:
> Hi @binary-koan 👋 Requesting a review for this PR. Description has been updated with all the details and demos. I have refactored the UI a bit today to be in line with the overall helper theme and have addressed your feedback above.

Hey @binary-koan, gentle ping for this. Let me know if there is something I have to address before you want to take a look

binary-koan commented:
Sorry for the delay! Will get to this today or tomorrow

binary-koan commented on app/(dashboard)/[category]/conversation/conversation.tsx:
```diff
@@ -411,7 +410,8 @@ const ConversationContent = () => {
   const { isAboveSm } = useBreakpoint("sm");
 
-  const defaultSize = Number(localStorage.getItem("conversationHeightRange") ?? 65);
+  const defaultSize =
+    typeof window !== "undefined" ? Number(localStorage.getItem("conversationHeightRange") ?? 65) : 65;
```
Are these changes related?

chuporustum replied:
Removed.

binary-koan commented on app/(dashboard)/[category]/conversation/issueAssignButton.tsx:
```diff
@@ -0,0 +1,96 @@
+  const { data: issueGroups } = api.mailbox.issueGroups.listAll.useQuery();
+  const utils = api.useUtils();
+
+  const assignMutation = api.mailbox.issueGroups.assignConversation.useMutation({
```
I think we should allow creating issues from here as well to avoid breaking people's flow with a jump to settings

binary-koan commented on app/(dashboard)/[category]/shared/queries.ts:
```diff
@@ -18,6 +18,7 @@ export const useConversationsListInput = () => {
     isPrompt: parseAsBoolean,
     reactionType: parseAsStringEnum(["thumbs-up", "thumbs-down"] as const),
     events: parseAsArrayOf(parseAsStringEnum(["request_human_support", "resolved_by_ai"] as const)),
+    issueGroupId: parseAsInteger,
```
I don't think this is currently a visible filter? We should show it in the filter bar IMO - it would both make the UI clearer and add an alternative (consistent) way to see issues in a group

binary-koan commented on app/(dashboard)/common-issues/page.tsx:
```diff
@@ -0,0 +1,323 @@
+    return filtered;
+  }, [data?.groups, searchQuery, sortBy]);
+
+  const handlePinGroup = (groupId: number, _cleanTitle: string) => {
```
```suggestion
  const handlePinGroup = (groupId: number) => {
```

Unused parameter?

binary-koan commented on app/(dashboard)/common-issues/page.tsx:
```diff
@@ -0,0 +1,323 @@
+            <>
+              <div className="grid gap-4 md:grid-cols-2 lg:grid-cols-3 items-stretch">
+                {filteredAndSortedGroups.map((group, index) => {
+                  const isPinned = pinnedData?.groups.some((p) => p.id === group.id) ?? false;
```
Would be nice to split out an item component here, this is a chunky file

binary-koan commented on app/(dashboard)/common-issues/page.tsx:
```diff
@@ -0,0 +1,323 @@
+                        <CardContent className="pt-0">
+                          <div className="flex items-center justify-between gap-3">
+                            <div className="flex items-center gap-2 flex-wrap">
+                              {(() => {
```
I think having components for these too rather than inline functions would be easier to read

binary-koan commented on db/schema/issueGroups.ts:
```diff
@@ -0,0 +1,30 @@
+  {
+    ...withTimestamps,
+    id: bigint({ mode: "number" }).primaryKey().generatedByDefaultAsIdentity(),
+    mailboxId: bigint("mailbox_id", { mode: "number" }).notNull(),
```
We don't need this column since we no longer support multiple mailboxes

binary-koan commented on jobs/publishIssueGroupEvent.ts:
```diff
@@ -0,0 +1,31 @@
+import { issueGroupsChannelId } from "@/lib/realtime/channels";
+import { publishToRealtime } from "@/lib/realtime/publish";
+
+export const publishIssueGroupEvent = async ({
```
This doesn't seem to be used?

binary-koan commented on jobs/handleIssueGroupStatusChange.ts:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  // Publish realtime update to refresh issue group counts
+  await publishToRealtime({
```
I'm not convinced we need this to update in real-time - it's a nice feature but adds complexity and potentially refetches the entire query a lot of times if a lot of messages are coming in

If we keep this around I think we should send the new count for this group in the event and use `setData` rather than `refetch` on the frontend, but I'd be find with just removing it

binary-koan commented on jobs/categorizeConversationToIssueGroup.ts:
```diff
@@ -0,0 +1,189 @@
+  return contentParts.join(" ");
+};
+
+const categorizeWithAI = async (
```
This job and prompt look great!

binary-koan commented on lib/data/conversation.ts:
```diff
@@ -134,6 +134,11 @@ export const updateConversation = async (
+
+  // Trigger issue group updates when conversation status changes
+  if (current.status !== updatedConversation.status) {
+    await triggerEvent("issue-groups/status-changed", { conversationId: updatedConversation.id });
```
If we keep this, it would also need to fire when conversation issue group is updated (either in the UI or by the job) as well, right?

binary-koan commented on trpc/router/mailbox/issueGroups.ts:
```diff
@@ -0,0 +1,371 @@
+import { mailboxProcedure } from "./procedure";
+
+export const issueGroupsRouter = {
+  // List all issue groups for a mailbox with conversation counts
```
I don't think we need these comments

binary-koan commented on trpc/router/mailbox/issueGroups.ts:
```diff
@@ -0,0 +1,371 @@
+      }));
+
+      // Filter out groups with 0 open conversations for the main view
+      const activeGroups = groups.filter((group) => group.openCount > 0);
```
I think we should still show groups with no tickets - you may want to view closed tickets or pin a new group which hasn't had anything come in yet

binary-koan commented on trpc/router/mailbox/issueGroups.ts:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    // Unassign all conversations from this group
+    if (group.conversations.length > 0) {
+      await db.update(conversations).set({ issueGroupId: null }).where(eq(conversations.issueGroupId, input.id));
+    }
+
+    // Delete the group
+    await db.delete(issueGroups).where(eq(issueGroups.id, input.id));
```
Would be nice to do this in a transaction

binary-koan reviewed (commented):
Sorry again for the long wait! Added some comments on the code; I haven't looked too much at the design but if you have Figma access now could you check against the mockup again?

lmk if you need clarification on anything!

binary-koan reviewed (commented):
Just been trying this out locally, and made a few design changes (slightly different middle ground between Figma and current conventions, so it was easier to just push changes than try to explain!)

I found a few UX issues while testing, could you look into them? Also added a couple more minor comments

* When you filter by an issue you've pinned, the sidebar item should show as active (and ideally "all" should be inactive)
* On the main page, the whole card should link to the conversation list not just the title
* Looks like you can't edit issues on the settings page?

chuporustum commented:
All comments have been addressed. cc @binary-koan 
Please checkout the new state via these recordings.

**Admin UI**

https://github.com/user-attachments/assets/a980c7de-92c1-4f96-a450-67560bf8600b

**Common Issues UI**

https://github.com/user-attachments/assets/cf91c3ac-bef8-4f84-949a-b77f871c7d64

chuporustum commented:
Hi @binary-koan 👋 
Gentle ping for this! This is open from quite some time now so just want to take this towards closure.
